<h2>Why is this an issue?</h2>
<p>This rule detects usage of APIs marked as deprecated in your project's configuration.</p>

<h2>Configuration</h2>
<p>Configure the deprecated APIs in the <code>deprecatedApis</code> rule parameter as a JSON array of objects.</p>

<h3>JSON Format Specification</h3>
<p>Each element in the array must contain the following fields:</p>
<table>
  <tr>
    <th>Field</th>
    <th>Type</th>
    <th>Required</th>
    <th>Description</th>
  </tr>
  <tr>
    <td><code>fqcn</code></td>
    <td>string</td>
    <td>Yes</td>
    <td>Fully qualified class name (e.g., <code>com.example.OldApi</code>)</td>
  </tr>
  <tr>
    <td><code>member</code></td>
    <td>string</td>
    <td>Yes</td>
    <td>Method or field name. Use <code>&lt;init&gt;</code> for constructors</td>
  </tr>
  <tr>
    <td><code>arguments</code></td>
    <td>string</td>
    <td>No</td>
    <td>Source-style argument list (e.g., <code>(java.lang.String)</code>). When null or omitted, all overloads are targeted</td>
  </tr>
  <tr>
    <td><code>migration</code></td>
    <td>string</td>
    <td>Yes</td>
    <td>Recommended action or replacement API</td>
  </tr>
  <tr>
    <td><code>note</code></td>
    <td>string</td>
    <td>No</td>
    <td>Additional context or version information</td>
  </tr>
</table>

<h3>Arguments Format</h3>
<p>The <code>arguments</code> field uses source-style notation:</p>
<ul>
  <li><code>()</code> for methods or constructors with no parameters</li>
  <li><code>(java.lang.String,int)</code> for multiple parameters</li>
  <li><code>(java.lang.String[])</code> for arrays or varargs</li>
</ul>

<h3>Example Configuration</h3>
<pre>
[
  {
    "fqcn": "com.example.OldApi",
    "member": "oldMethod",
    "arguments": null,
    "migration": "Use com.example.NewApi#newMethod() instead",
    "note": "Will be removed in v3.0"
  },
  {
    "fqcn": "com.example.Constants",
    "member": "OLD_CONSTANT",
    "arguments": null,
    "migration": "Use Constants.NEW_CONSTANT",
    "note": ""
  },
  {
    "fqcn": "com.example.Api",
    "member": "process",
    "arguments": "(java.lang.String)",
    "migration": "Use processNew(String) instead",
    "note": "Only the String variant is deprecated"
  },
  {
    "fqcn": "com.example.Builder",
    "member": "&lt;init&gt;",
    "arguments": "(java.lang.String,java.lang.String)",
    "migration": "Use the builder pattern instead",
    "note": "Specific constructor arguments"
  }
]
</pre>

<h2>Noncompliant Code Example</h2>
<pre>
OldApi api = new OldApi();
api.oldMethod(); // Noncompliant
</pre>

<h2>Compliant Solution</h2>
<pre>
NewApi api = new NewApi();
api.newMethod();
</pre>
